### ISSUE TITLE
Making POST request on class or Collection instance returns Internal Server Error  

### ISSUE THREAD

**Issue by priyanshunayan** [HTTP-APIs/hydrus#549]
### I'm submitting a 
- [x] bug report.
- [ ] feature request.

### Current Behaviour:

When making a POST or DELETE request to the class instance with a new request body gives an error. But it does update the instance with new values. 

### Expected Behaviour:

It should not give the error.

### Steps to reproduce:

1. Start hydrus server (master revision) with this APIDoc: https://gist.github.com/priyanshunayan/0dc02d65e1f036b9a45919b5342d00a4
2. Create a new class instance by making PUT to `/Movie` endpoint with request body:
```json
{
    "@type": "Movie",
    "movie_name": "Godfather",
    "movie_director": "F.F Coppola"
}
```
3. Now make a POST request to this class instance at `/Movie/object-id` with updated request body
4. It should return the following error:
```bash

Traceback (most recent call last):
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask_restful/__init__.py", line 272, in error_router
    return original_handler(e)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/_compat.py", line 38, in reraise
    raise value.with_traceback(tb)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask_restful/__init__.py", line 272, in error_router
    return original_handler(e)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/_compat.py", line 38, in reraise
    raise value.with_traceback(tb)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask_restful/__init__.py", line 468, in wrapper
    resp = resource(*args, **kwargs)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask/views.py", line 89, in view
    return self.dispatch_request(*args, **kwargs)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask_restful/__init__.py", line 583, in dispatch_request
    resp = meth(*args, **kwargs)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/hydrus/auth.py", line 89, in wrapper
    return f(*args, **kwargs)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/hydrus/resources.py", line 123, in post
    return items_post_check_support(id_, object_, class_path, path, is_collection)
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/hydrus/itemhelpers.py", line 84, in items_post_check_support
    send_sync_update(socketio=socketio, new_job_id=new_job_id,
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/hydrus/helpers.py", line 367, in send_sync_update
    socketio.emit('update', data, namespace="/sync")
  File "/Users/priyanshunayan/Developer/creating_api_doc/venv/lib/python3.8/site-packages/flask_socketio/__init__.py", line 416, in emit
    self.server.emit(event, *args, namespace=namespace, room=room,
AttributeError: 'NoneType' object has no attribute 'emit'

-->
```

### Snapshot:

![image](https://user-images.githubusercontent.com/32892545/105709456-bf2c6900-5f3b-11eb-9b8e-8bff6ce39f0b.png)

### Environment:
 
* python version - 3.8.2
* pip version - pip 20.1.1
* OS details - macOS Big Sur(11.1)

### Do you want to work on this issue?

no

**Comment by priyanshunayan** [comment]
It happens with DELETE operation as well.

**Comment by Mec-iS** [comment]
thanks for the bug report.
just a note to avoid overflow of information: there is no need to paste the entire HTML code, the Python stacktrace and the screenshot are enough.

It seems that the helper is calling from the wrong object when it does `self.server.emit()`. `self.server` is not defined. @sameshl

**Comment by Purvanshsingh** [comment]
![Screenshot from 2021-01-28 21-14-56](https://user-images.githubusercontent.com/49719371/106164228-c22f8f80-61af-11eb-8686-8025c7602146.png)
We are directly using `socketio` instead of using `create_socket` method from **socketio_factory.py**,
by default, the value of self.server is None that's why we are getting this error in POST & DELETE operation.
I will create the PR of the fix.

**Comment by farazkhanfk7** [comment]
I don't think we need any fix for this. Can you please confirm that you're using `socketio.run(app, port=port)` to run the server in your python file ? 
Hydrus runs the server using cli.py. I tried it after changing apidoc_path to sample Movie apidoc's path in conf.py and it works fine with classes. Tried making POST request to update Collections and it simply returns a 405 status code (Method not allowed) as expected. DELETE operation also works fine.

**Comment by Purvanshsingh** [comment]
![Screenshot from 2021-01-29 01-17-06](https://user-images.githubusercontent.com/49719371/106190710-0894e680-61d0-11eb-8bdd-96bf539ed11c.png)

**Comment by farazkhanfk7** [comment]
I think you should use create_socket that takes an instance of app and session like `socketio = create_socket(app, session)`
See if this could solve the issue : https://gist.github.com/farazkhanfk7/d2a932c85f952739671a9d4fda60c5da
This should work.

**Comment by Purvanshsingh** [comment]
@priyanshunayan this issue is still open, resolved with #552
